### ISSUE TITLE
Add traps after semantic actions to prevent undefined control flow.

### ISSUE THREAD

**Issue by skvadrik** [skvadrik/re2c#305]
User-defined semantic actions should have a control flow transfer statement at the end (one of `goto`, `return`, `continue`, `break`, etc). If the user forgets to add one, the program may fail in an unexpected way. This is especially confusing for new users.

The suggestion is to add some sort of `assert(false)` or `__builtin_trap()` statement after the user-defined semantic action. It should be unreachable (dead code) in correct programs, so some way of suppressing compiler warnings is needed.

**Comment by DemiMarie** [comment]
What about a call to an undefined function?  That will cause a link failure if not optimized away.  GCC has an attribute that can be used to give a better message.

**Comment by skvadrik** [comment]
> What about a call to an undefined function?

This is a nice hack, but re2c should not rely on the compiler optimizations: it can be TCC or GCC/Clang without optimizations.

**Comment by DemiMarie** [comment]
> > What about a call to an undefined function?
> 
> This is a nice hack, but re2c should not rely on the compiler optimizations: it can be TCC or GCC/Clang without optimizations.

True, though with GCC there is a preprocessor macro that can be used to check if optimizations are on.

**Comment by skvadrik** [comment]
Closing this, as there is a way to add traps via ``!post_rule`` in whatever way is most convenient for the user. Please reopen if you think there's something else to be done here.

**Comment by ArtemPisarenko** [comment]
> (jumping to the end of lexer would be good, but it may add unreachable instructions and cause compiler warnings/errors)

It is indeed good (at least for C/C++). And such behavior of compiler enforces user to account this when `re2c:cond:abort` isn't enabled (default for C/C++). It's **not** bad that compiler complains when it actually finds that control flow reaches end of non-void function body without return statement. It's **bad** that user didn't write this statement naively relying on expectation that control flow from generated block will never reach it. If user relies on that then it must be sure that all rules/actions and semantic actions defined such way, and compiler warning/error will say "no, your expectation is wrong".

I'm sure it's a user responsibility to handle reaching end of block (and other kinds of not-handled control flows) unless user **explicitly** delegated this responsibility to re2c by `re2c:cond:abort = 1`.

**Comment by ArtemPisarenko** [comment]
Also it would be good in `--conditions` mode to make `=>` syntax behavior be extended version of `:=>` by adding transition right after semantic action. It would be intuitively expected and preventing undefined control flow. I've got into trap when initially written rule which only transitions to condition and then decided to add semantic action expecting that it will be executed before transition, but removing `:` also removed transition.
Yes, it's my fault, I overlooked it when read docs. But it happened because I didn't expected that now I have to duplicate things by writing

- either `<...> regex => Condition { pre-action; goto yyc_Condition;}`
- or `<...> regex { yystate = yycCondition; pre-action; goto yyc_Condition;}`

(where `pre-action` is rule-specific usercode, not something which might be delegated to `!pre_rule` syntax).

Going further it would be nice also to avoid somewhat similar duplication demonstrated by function `parse_u32(const char *s)` in example from official docs. That is when we want to proceed to same condition start after matching rule which itself doesn't involve any jumping (either from semantic action, or by omitting `=>` / `:=>`). Similar case is when it's a single common rule for condition-list (in this case we have to write dispatcher in semantic action). I don't think that it may be solved by placing jumping to condition start after semantic action of such "non-transition" rule, because it will be inconsistent with non-condition re2c mode (where we want to jump at block end?). I would prefer to introduce additional syntax for this kind of rule. Something like `<...> regex <= { pre-action; } // jumps back to same condition start label as it done by initial dispatcher (but skipping !entry action?)`.
